### PR TITLE
Production-ready circuits

### PR DESCRIPTION
**Pull request by weijiekoh**
This PR introduces the `small` circuits which support the following tree capacities:

State tree: 2 ** 8 == 256 (maximum 255 users as the 0th leaf is random)
Message tree: 2 ** 11 = 2048 (maximum 2048 messages)
Vote options tree: 5 ** 3 = 125 (maximum 125 vote options)
State leaf batch size: 4
Message batch size: 4

If there are 2048 messages, we need 512 `batchProcessMessage()` invocations to process all messages. Additionally if there are 255 users, we need 64 `proveVoteTallyBatch()` invocations to tally all votes. As such, it is best to use a sidechain or L2 solution to reduce transaction costs.

I did some benchmarks on my laptop to experiment with different circuit sizes. I found that:

- my laptop has just enough memory to build the above `small` configuration
- the batch size has a big impact on the amount of memory needed
- even if I can build the circuit when the batch size is 8, I run into the following WASM error: `CompileError: AsyncCompile: Wasm decoding failed: data segments count of 116336 exceeds internal limit of 100000 @+1150381`

To try out the `small` circuits, see the "Demonstration with `prod-small` settings" section of `cli/README.md`.

The following tables are for reference only:

**`BatchUpdateStateTree`**

Tree depths vs circuit build time:

| State tree | Message tree | Vote options tree | Message batch size | `max-old-space-size` | `stack-size` | Circuit build time | Constraints |
|-|-|-|-|-|-|-|-|
| 7 | 10 | 3 | 8 | 16384 | 16384 | 12 mins | 231115 |
| 7 | 11 | 3 | 8 | 16384 | 16384 | 8 mins | 233019 |
| 8 | 11 | 3 | 8 | 16384 | 16384 | 8 mins | 237049 |
| 8 | 11 | 3 | 16 | 16384 | 16384 | OOM |  |

Tree depths vs circuit build time:

**`QuadVoteTally`**

| State tree | Intermediate state tree |  Vote options tree | `max-old-space-size` | `stack-size` | Circuit build time | Constraints |
|-|-|-|-|-|-|-|
| 7 | 3 | 3 | 16384 | 16384 | 7 mins  | 122404 |

### CODE REVIEW

File: cli/README.md [hunk 1]
```diff
@@ -489,3 +489,72 @@ The total spent voice credit commitment in the MACI contract on-chain is valid
...
+	-p macipk.40270618e1797c4969587eb04d7f3e9b39a91ecbbdf7d3c998d8e34d08e11c86 \
+	-i 1 \
+	-v 0 \
+	-w 9 \
+	-n 1
+```
+
+Process all messages:
+
+Remember to set `NODE_ENV=prod-small`!
```
Comment (xuhcc) [inline cli/README.md:533]: Would it be possible to run commands without setting NODE_ENV?

File: cli/README.md [hunk 2]
```diff
@@ -489,3 +489,72 @@ The total spent voice credit commitment in the MACI contract on-chain is valid
...
+## Demonstration with `prod-small` settings
+
+Create a MACI instance:
+
+```bash
+node ./build/index.js create -d 0xc87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3 \
+	-sk macisk.8715ab59a3e88a7ceec80f214ec24a95287ef2cb399a329b6964a87f85cf51c \
+	-e http://localhost:8545 \
+    -u 255
+    -m 2048
```
Comment (xuhcc) [inline cli/README.md:502]: It seems that `-u`, `-m` and `-v` parameters are no longer used by the CLI

File: cli/README.md (hunk 1 above)
Comment (weijiekoh) [reply cli/README.md:533]: Yup! I'll make some changes to support this.

File: cli/README.md (hunk 2 above)
Comment (weijiekoh) [reply cli/README.md:502]: They should be. I'll check.

Comment (xuhcc) [reply cli/README.md:502]: Ah sorry, you're right. 

They're just handled differently now https://github.com/appliedzkp/maci/pull/165/files#diff-6ef4b7073606e835cb72390d47842436L231-L235

**Comment by weijiekoh** [comment]
Pushed 0.2.2 to NPM.

**Comment by xuhcc** [comment]
@weijiekoh 

I tried to run voting using (8, 11, 3) parameter set, but processing of messages fails: `VM Exception while processing transaction: revert MACI: invalid batch UST proof`.

**Comment by weijiekoh** [comment]
> @weijiekoh
> 
> I tried to run voting using (8, 11, 3) parameter set, but processing of messages fails: `VM Exception while processing transaction: revert MACI: invalid batch UST proof`.

Fixing this in https://github.com/appliedzkp/maci/pull/166

**Comment by xuhcc** [comment]
I'm getting the same error (`MACI: invalid batch UST proof`) using MACI 0.2.3 and (4, 4, 2) tree depths, but tests are now passing with (8, 11, 3).

- (4, 4, 2): `BatchUpdateStateTreeVerifier` and `QuadVoteTallyVerifier` contracts.
- (8, 11, 3): `BatchUpdateStateTreeVerifierSmall` and `QuadVoteTallyVerifierSmall` contracts.

**Comment by xuhcc** [comment]
Contracts `BatchUpdateStateTreeVerifier.sol` and `BatchUpdateStateTreeVerifierSmall.sol` are identical, shouldn't they be different?

**Comment by weijiekoh** [comment]
Oops. I'll fix this in #167 .

**Pull request merged**